### PR TITLE
resolves MRO error in MutualAuth class in Python3

### PR DESCRIPTION
**Pull request by ehashman**
Helping @fboxwala open this PR since she said she had a permissions problem.

@lvh, I had switched the order of these classes per your instruction; this PR would essentially revert that (complete commit in question: https://github.com/RackSec/overcorps/pull/78/commits/3bdf689ce481b9779bf6055a18a191fdd826a0f4)

@fboxwala, can you give more detail on why you want to do this?

### CODE REVIEW

**Comment by fboxwala** [comment]
When I run `tox -e py35` on account-service I get the following error with the current ordering:

```
tests/test_api.py:3: in <module>
    import account_service.api
account_service/api.py:3: in <module>
    from falcon_mutualauth.mutualauth import Authorize
.tox/py35/lib/python3.5/site-packages/falcon_mutualauth/mutualauth.py:24: in <module>
    class MutualAuthRequest(object, Request):
E   TypeError: Cannot create a consistent method resolution
E   order (MRO) for bases Request, object
```

When I switch object and Request I don't get the error, @reaperhulk helped me figure this out earlier today

Switching the order doesn't affect the tests in py27 - so we figured it would not break anything in overcorps

**Comment by lvh** [comment]
The underlying issue here is that `Request` is an old-style class on Python 2, and a new-style class in Python 3 (where old-style classes do not exist). Those two have different method resolution schemes. If we can live with `Request` staying old-style on Python 2, we can just remove `object`. Reverting the patch would reintroduce the underlying bug on Python 2.x, which IIUC is still the primary platform we care about here.

I think not making Request new-style is the best option, although that does mean that e.g. we can't put properties on it.

**Comment by reaperhulk** [comment]
Out of curiosity, what was the bug in 2.x?

**Comment by lvh** [comment]
The bug in 2.x is that an old-style class was next-in-line for the MRO and being called by `super`. I didn't write the code to be 3.x compatible because two point nine overbar eleventy. If we do actually care about 3.x compat, and we remove object, we should also remove the `super()` call which will, um, sorta make it work (ideally, you'd use six's PY3 detection to do either the super or the explicit `__init__`).

shitty transcript from slack privmsg:

---

lvh [4:15 PM]  
so, the short version is that it relies on the next class in the MRO accidentally being the bottom of the inheritance diamond; because that diamond is resolved by explicit **init**, and not by super

[4:15]  
so, injecting super in it can, for example, call **init** more than once

Paul Kehrer [4:15 PM]  
nasty

lvh [4:16 PM]  
by putting object there, you are terminating the chain at object because object stops calling **init**

[4:16]  
So solutions include: don’t retroactively try to make a new-style Request class in 2.x, or stop using super

[4:17]  
but you can’t stop using super and enforce correctness without enforcing that your subclasses ​_also_​ don’t use super or don’t have inheritance diamonds

Paul Kehrer [4:17 PM]  
but in both cases here object is in the inheritance hierarchy. the code before had object listed first and the change lists it last

lvh [4:17 PM]  
yes; but that changes the MRO

Paul Kehrer [4:18 PM]  
wouldn't the change make object be the bottom where previously it would look on object first?

lvh [4:18 PM]  
Part of the problem is with object being present at all, a different part of the problem is if it’s present, when does it participate in the MRO

[4:18]  
which change? there’s two going in opposite directions :smile:

Paul Kehrer [4:19 PM]  
haha, well the PR changes it from MutualAuth(object, Request) to MutualAuth(Request, object)

lvh [4:20 PM]  
putting Request first puts Request earlier in the MRO here

[4:20]  
by putting Request last in the MRO, object terminates the super chain, so you can use super

[4:21]  
you then ​_also_​ call Request.**init** to correctly kick off the **init** old style chain

Paul Kehrer [4:23 PM]  
okay, I follow :slightly_smiling_face:

lvh [4:24 PM]  
as to why that’s a problem, I present exhibit A of request being a little sad

[4:24]  

```
from twisted.web.server import Request; type("", [Request, object], {}).__mro__
Out[11]:
(__main__.,
<class twisted.web.server.Request at 0x000000010acc7ee8>,
<class twisted.spread.flavors.Copyable at 0x000000010ab71b00>,
<class twisted.spread.flavors.Serializable at 0x000000010ab712b8>,
<class twisted.spread.jelly.Jellyable at 0x000000010ab71218>,
<class twisted.web.http.Request at 0x000000010ac304a8>,
<class twisted.python.components.Componentized at 0x000000010aa75ba0>,
object)
```

Paul Kehrer [4:24 PM]  
ha, that I did look up in the docs

[4:24]  
multiple inheritance out the ass

lvh [4:24 PM]  
yup.

[4:25]  
So, basically, this affects people who want to compose, using the best tool available for composition: inheritance

[4:25]  
specifically if they want inherit from this class and their own wonky Request class)

---

**Comment by fboxwala** [comment]
So I should not change this in mutual-auth. Does that mean I should make account-service not be py3 compatible?

**Comment by ehashman** [comment]
@fboxwala: no, I'd say your best bet here is to switch this back to an old-style class (i.e. don't inherit from object), instead of trying to mess with the MRO/changing order of multiple inheritance. I believe this is what @lvh suggested.

**Comment by lvh** [comment]
Yep, although caveat emptor: to make that correct, you _must_ also remove the `super` call.

**Comment by lvh** [comment]
Looking at the implementation, that appears to mean you can just remove `__init__` entirely.

**Comment by reaperhulk** [comment]
LGTM
